x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+                        public void onResponse(BulkResponse response) {
+                            for (BulkItemResponse itemResponse : response.getItems()) {
+                                if (itemResponse.isFailed() == false) {
+                                    assert itemResponse.getResponse() instanceof DeleteResponse;
+                                    maintenanceTask.deletes.incrementAndGet();
+                                }
+                            }
+                            maintenanceTask.searchResponse = null;
+                            maintenanceTask.searchAfter = finalSearchAfter;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+                // we're done, complete the task
+                complete(this, null);
+            } catch (Exception e) {
+                complete(this, e);
+            }
+        }
+
+        public boolean isClosed() {
+            return closed.get();
+        }
+
+        private void ensureOpen() {
+            if (isClosed()) {
+                assert false : "should not use periodic task after close";
+                throw new IllegalStateException("Periodic maintenance task is closed");
+            }
+        }
+
+        @Override
+        public void close() {
+            if (closed.compareAndSet(false, true)) {
+                final Exception e = error.get();
+                if (e != null) {
+                    logger.warn(
+                        () -> new ParameterizedMessage(
+                            "periodic maintenance task completed with failure ({} deleted documents out of a total of {})",
+                            deletes.get(),
+                            total.get()
+                        ),
+                        e
+                    );
+                } else {
+                    logger.info(
+                        () -> new ParameterizedMessage(
+                            "periodic maintenance task completed ({} deleted documents out of a total of {})",
+                            deletes.get(),
+                            total.get()
+                        )
+                    );
+                }
+            }
+        }
+
+        private boolean assertGenericThread() {
+            final String threadName = Thread.currentThread().getName();
+            assert threadName.contains(ThreadPool.Names.GENERIC) : threadName;
+            return true;
+        }
+    }
+
+    private void executeNext(PeriodicMaintenanceTask maintenanceTask) {
+        threadPool.generic().execute(maintenanceTask);
+    }
+
+    private void complete(PeriodicMaintenanceTask maintenanceTask, @Nullable Exception failure) {
```

Comment:
Could this not be a method on PeriodicMaintenanceTask to avoid the `maintenanceTask` parameter here and the dereferences of it in the implementation?

Reply:
Sure, I pushed b009c35094d88308c6284012f9d0a760dc1e95cb.